src/irmin-pack/unix/file_manager.ml

```diff
@@ -103,64 +100,66 @@ struct
   (** Flush stage 1 *)
   let flush_dict t =
     let open Result_syntax in
-    if Dict.empty_buffer t.dict then Ok ()
-    else
-      let* () =
+    let* () =
+      if Dict.empty_buffer t.dict then Ok ()
+      else (
         Stats.incr_fm_field Dict_flushes;
-        Dict.flush t.dict
-      in
-      let* () = if t.use_fsync then Dict.fsync t.dict else Ok () in
-      let* () =
-        let pl : Payload.t = Control.payload t.control in
-        let pl = { pl with dict_end_poff = Dict.end_poff t.dict } in
-        Control.set_payload t.control pl
-      in
-      let+ () = if t.use_fsync then Control.fsync t.control else Ok () in
-      ()
+        Dict.flush t.dict)
+    in
+    if t.use_fsync then Dict.fsync t.dict else Ok ()
```

Comment:
Subtle change here: the previous code did not fysnc if the buffer is empty. I think we should preserve that.

Reply:
Ha I remember there was some subtlety here... The dict could have already dumped its buffer to disk earlier, so an empty buffer is not indicative of whether we should flush or not. I'm going to fix this by keeping a flag in the append-only file to remember if an fsync is required :)